File: JavaScript/127-Word Ladder.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLength = function(beginWord, endWord, wordList) {
    let wordSet = new Set(wordList);
    if (!wordSet.has(endWord)) return 0;
    if (wordSet.has(beginWord)) wordSet.delete(beginWord);
    
    let queue1 = new Set([beginWord]), queue2 = new Set([endWord]);
    
    let step = 1
    while (queue1.size && queue2.size){
        if (queue1.size > queue2.size){
            [queue1, queue2] = [queue2, queue1];
        }
        let tmp = new Set();
        for (let currWord of queue1){
            for (let i=0; i<currWord.length; i++){
                for (let char of "abcdefghijklmnopqrstuvwxyz"){
                    let newWord = currWord.slice(0,i) + char +currWord.slice(i+1);
                    if (queue2.has(newWord)){
                        return step+1;
                    }
                    if (wordSet.has(newWord)){
                        wordSet.delete(newWord);
                        tmp.add(newWord);
                    }
                }
            }
        }
        queue1 = tmp;
        step += 1;
    }
    return 0;  
};

File: 127-Word Ladder.py
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
# BFS, O(M*M*N), M:length of word, N: length of wordList
        wordSet = set(wordList)
        
        if beginWord in wordList:
            wordSet.remove(beginWord)
        
        queue = collections.deque([beginWord])
        
        step = 1
        
        while queue:
            for _ in range(len(queue)):
                currWord = queue.popleft()
                if currWord == endWord:
                    return step
                for i in range(len(currWord)):
                    for char in "abcdefghijklmnopqrstuvwxyz":
                        newWord = currWord[:i] + char + currWord[i+1:]
                        if newWord in wordSet:
                            wordSet.remove(newWord)
                            queue.append(newWord)
            step += 1
        return 0
    
# bidirectional BFS
        wordSet = set(wordList)
        
        if beginWord in wordList:
            wordSet.remove(beginWord)
        if endWord not in wordList:
            return 0
        
        queue1 = set([beginWord])
        queue2 = set([endWord])
        
        step = 1
        
        while queue1 and queue2:
            if len(queue1) > len(queue2):
                queue1, queue2 = queue2, queue1
            tmp = set()
            for currWord in queue1:
                for i in range(len(currWord)):
                    for char in "abcdefghijklmnopqrstuvwxyz":
                        newWord = currWord[:i] + char + currWord[i+1:]
                        if newWord in queue2:
                            return step+1
                        if newWord in wordSet:
                            wordSet.remove(newWord)
                            tmp.add(newWord)
            queue1 = tmp
            step += 1
            
        return 0